proposals/NNNN-unsafe-task-locals.md

```diff
@@ -0,0 +1,202 @@
+Structured Concurrency by forcing the APIs to form scopes during which the value binding applies.
+
+In practice this means that there is no API to "set" a task-local value, but instead a key is bound to a specific value
+in the task's storage by calling the `withValue` method, like this:
+
+```swift
+@TaskLocal
+var currentCustomer: CustomerID?
+
+let customerID = makeCustomerID()
+
+$currentCustomer.withValue(customerID) {
+    let sushi = makeSushi()
+    // ... 
+}
+
+func makeSushi() -> Sushi {
+    print("Preparing sushi for customer \(currentCustomer)")
+    // ... 
+}
+```
+
+This capability is used primarily to propagate additional metadata throughout a chain of calls, 
+without which the system still would work fine, but thanks to this metadata we can provide
+enriched logging, security, or other semantics that improve the usability and understandability of the system.
+
+Notably, the `withValue` closure makes it possible for the runtime to reliably "push" and "pop" values onto the 
+task-local bindings stack. Developers cannot "forget" to remove a value when it should be cleaned up, and they cannot
+remove more values than they should. This API is inherently safe and does not need much explanation to use it correctly.
+
+This API however does not compose well when code being called cannot be wrapped in a closure like this.
+And more specifically, it makes writing method attached macros problematic because they would have to "replace"
+the entire method body, sacrificing performance, safety, and auto-completion in the process.
+
+This proposal offers an unsafe solution tailored to advanced users to remedy this problem.
+
+## Proposed solution
+
+We propose to introduce a pair of "unsafe" methods used to push and pop value bindings onto the task-local bindings stack:
+
+- `unsafePushValue(_:)`
+- `unsafePopValue()`
+
+They are equivalent to what the `withValue` method is currently doing, however they can be used without forcing the 
+nesting that the with closure accepting API forces (`$key.withValue(...) { ... }`).
+
+These APIs are unsafe because an un-balanced pop could either attempt to pop more values than the stack contains 
+leading to incorrect value bindings. Or even worse attempt to pop a value stored in a parent task (which can happen
+by popping more values than a task has pushed). 
+
+> **WARNING:** Mis-use of these APIs will lead to incorrect behavior, crashes, or even undefined behavior, so they must be handled very carefully. Preferably, only in macro generated code which can _guarantee_ the correct balance of push/pop calls.
+
+The following code using `withValue`:
+
+```swift
+func order() async {
+  $wasabiPreference.withValue(.with) {
+    let sushi = await orderSushi()
+  }
+}
+```
+
+can be expressed using the new unsafe APIs as follows:
+
+```swift
+func order() async {
+  $wasabiPreference.unsafePushValue(.with)
+  defer { TL.$number.unsafePopValue() } 
```

Comment:
Where is `TL.$number` coming from? Shouldn't this be `$wasabiPreference.unsafePopValue()`?